index.js

```diff
@@ -702,101 +709,113 @@ HardSourceWebpackPlugin.prototype.apply = function(compiler) {
-          loaders: module.loaders,
-          identifier: module.identifier(),
-          // libIdent: module.libIdent &&
-          // module.libIdent({context: compiler.options.context}),
-          assets: Object.keys(module.assets || {}),
-          buildTimestamp: module.buildTimestamp,
-          strict: module.strict,
-          meta: module.meta,
-          used: module.used,
-          usedExports: module.usedExports,
-
-          rawSource: module._source ? module._source.source() : null,
-          source: source.source(),
-          map: devtoolOptions && source.map(devtoolOptions),
-          // Some plugins (e.g. UglifyJs) set useSourceMap on a module. If that
-          // option is set we should always store some source map info and
-          // separating it from the normal devtool options may be necessary.
-          baseMap: module.useSourceMap && source.map(),
-          hashContent: serializeHashContent(module),
-
-          dependencies: serializeDependencies(module.dependencies),
-          variables: serializeVariables(module.variables),
-          blocks: serializeBlocks(module.blocks),
-
-          fileDependencies: module.fileDependencies,
-          contextDependencies: module.contextDependencies,
-
-          errors: module.errors.map(serializeError),
-          warnings: module.warnings.map(serializeError),
-        };
+        promise = md5(module.resource).then(function(md5Contents) {
+          if (
+            existingCacheItem &&
+            md5Contents !== existingCacheItem.md5Contents ||
+            !existingCacheItem
+          ) {
+            var source = module.source(
+              compilation.dependencyTemplates,
+              compilation.moduleTemplate.outputOptions,
+              compilation.moduleTemplate.requestShortener
+            );
+            var assets = Object.keys(module.assets || {}).map(function(key) {
+              return {
+                key: requestHash(key),
+                value: module.assets[key].source(),
+              };
+            });
+
+            moduleCache[identifier] = {
+              moduleId: module.id,
+              context: module.context,
+              request: module.request,
+              userRequest: module.userRequest,
+              rawRequest: module.rawRequest,
+              resource: module.resource,
+              loaders: module.loaders,
+              identifier: module.identifier(),
+              // libIdent: module.libIdent &&
+              // module.libIdent({context: compiler.options.context}),
+              assets: Object.keys(module.assets || {}),
+              buildTimestamp: module.buildTimestamp,
+              strict: module.strict,
+              meta: module.meta,
+              used: module.used,
+              usedExports: module.usedExports,
+
+              rawSource: module._source ? module._source.source() : null,
+              source: source.source(),
+
+              md5Contents: md5Contents,
```

Comment:
Don't store the md5 on the module. We need to store it in its own cache or in the data cache. We need an md5 for each entry in fileDependencies. Would be good to store mtime too to do determine if those stored md5's are reliable.